### PR TITLE
schemacrawler: init at 16.20.4

### PR DESCRIPTION
**Pull request by elohmeier**
## Description of changes

Added Schemacrawler.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/schemacrawler-bin/default.nix [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+, stdenv
+, fetchurl
+, unzip
+, jre
+}:
+
+stdenv.mkDerivation rec {
+  pname = "schemacrawler-bin";
+  version = "16.20.4";
+  src = fetchurl {
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler-bin/default.nix:11]: How about using `fetchzip`?

File: pkgs/development/tools/schemacrawler-bin/default.nix [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    url = "https://github.com/schemacrawler/SchemaCrawler/releases/download/v${version}/schemacrawler-${version}-bin.zip";
+    sha256 = "sha256-YozWe3JMCG+NtGZlDp2zSIV7oWaKaep0zs5Lfr/CSWU=";
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler-bin/default.nix:13]:
```suggestion
    hash = "sha256-YozWe3JMCG+NtGZlDp2zSIV7oWaKaep0zs5Lfr/CSWU=";
```

File: pkgs/development/tools/schemacrawler-bin/default.nix [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp -r {config,lib} $out/
+    cat >$out/bin/schemacrawler <<EOF
+    #!/bin/sh
+    ${jre}/bin/java -cp $out/lib/*:$out/config schemacrawler.Main "\$@"
+    EOF
+    chmod +x $out/bin/schemacrawler
+  '';
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler-bin/default.nix:26]: I think you can use `makeWrapper` here.

File: pkgs/development/tools/schemacrawler-bin/default.nix [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  preferLocalBuild = true;
+
+  meta = with lib; {
+    description = "Database schema discovery and comprehension tool";
+    homepage = "https://www.schemacrawler.com/";
+    license = licenses.gpl3;
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler-bin/default.nix:33]:
```suggestion
    license = with licenses; [ epl10 gpl3Only lgpl3Only ];
```
see https://github.com/schemacrawler/SchemaCrawler/blob/8c11b547c468f1b015010c363ca60066b677545b/schemacrawler-api/src/main/java/schemacrawler/Version.java#L13-L15

File: pkgs/development/tools/schemacrawler-bin/default.nix [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    ${jre}/bin/java -cp $out/lib/*:$out/config schemacrawler.Main "\$@"
+    EOF
+    chmod +x $out/bin/schemacrawler
+  '';
+
+  preferLocalBuild = true;
+
+  meta = with lib; {
+    description = "Database schema discovery and comprehension tool";
+    homepage = "https://www.schemacrawler.com/";
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler-bin/default.nix:32]:
```suggestion
    homepage = "https://www.schemacrawler.com/";
    sourceProvenance = with sourceTypes; [ binaryBytecode ];
```
Would you please set `sourceProvenance` and use `pname = "schemacrawler";` instead?
https://nixos.org/manual/nixpkgs/stable/#sec-meta-sourceProvenance

File: pkgs/development/tools/schemacrawler-bin/default.nix [hunk 6]
```diff
@@ -0,0 +1,37 @@
+{ lib
+, stdenv
+, fetchurl
+, unzip
+, jre
+}:
+
+stdenv.mkDerivation rec {
+  pname = "schemacrawler-bin";
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler-bin/default.nix:9]:
```suggestion
  pname = "schemacrawler";
```

File: pkgs/development/tools/schemacrawler-bin/default.nix [hunk 7]
```diff
@@ -0,0 +1,37 @@
+{ lib
+, stdenv
+, fetchurl
+, unzip
+, jre
+}:
+
+stdenv.mkDerivation rec {
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler-bin/default.nix:8]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
```
options: if you are interested
https://nixos.org/manual/nixpkgs/stable/#mkderivation-recursive-attributes

File: pkgs/development/tools/schemacrawler-bin/default.nix (hunk 1 above)
Comment (elohmeier) [reply pkgs/development/tools/schemacrawler-bin/default.nix]: Thanks, done!

File: pkgs/development/tools/schemacrawler-bin/default.nix (hunk 2 above)
Comment (elohmeier) [reply pkgs/development/tools/schemacrawler-bin/default.nix]: Done!

File: pkgs/development/tools/schemacrawler-bin/default.nix (hunk 3 above)
Comment (elohmeier) [reply pkgs/development/tools/schemacrawler-bin/default.nix]: Good idea, done!

File: pkgs/development/tools/schemacrawler-bin/default.nix (hunk 5 above)
Comment (elohmeier) [reply pkgs/development/tools/schemacrawler-bin/default.nix]: Added, thanks!

File: pkgs/development/tools/schemacrawler-bin/default.nix (hunk 6 above)
Comment (elohmeier) [reply pkgs/development/tools/schemacrawler-bin/default.nix]: Done

File: pkgs/development/tools/schemacrawler-bin/default.nix (hunk 7 above)
Comment (elohmeier) [reply pkgs/development/tools/schemacrawler-bin/default.nix]: Thanks, added!

File: pkgs/development/tools/schemacrawler/default.nix [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+  src = fetchzip {
+    url = "https://github.com/schemacrawler/SchemaCrawler/releases/download/v${finalAttrs.version}/schemacrawler-${finalAttrs.version}-bin.zip";
+    hash = "sha256-wNvFutxB34gjBCwVPIj/5jjjk0/IjH0EbD++FxxuweM=";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
+    mkdir -p $out/bin
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler/default.nix:20]:
```suggestion
  installPhase = ''
    runHook preInstall
  
    mkdir -p $out/bin
```

File: pkgs/development/tools/schemacrawler/default.nix [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    cp -r {config,lib} $out/
+
+    makeWrapper ${jre}/bin/java $out/bin/schemacrawler \
+      --add-flags "-cp $out/lib/*:$out/config" \
+      --add-flags schemacrawler.Main
+  '';
```
Comment (natsukium) [inline pkgs/development/tools/schemacrawler/default.nix:26]:
```suggestion
      --add-flags schemacrawler.Main
      
    runHook postInstall
  '';
```

**Review by natsukium — COMMENTED** [review]
Besides this, LGTM

File: pkgs/development/tools/schemacrawler/default.nix (hunk 8 above)
Comment (elohmeier) [reply pkgs/development/tools/schemacrawler/default.nix]: Thanks, fixed!

File: pkgs/development/tools/schemacrawler/default.nix (hunk 9 above)
Comment (elohmeier) [reply pkgs/development/tools/schemacrawler/default.nix]: Thanks, fixed!

**Review by natsukium — APPROVED** [review]
LGTM, thanks!

**Pull request merged**